bpf-gpl/parsing.h

```diff
@@ -19,7 +19,20 @@
 #define __CALI_PARSING_H__
 
 static CALI_BPF_INLINE int parse_packet_ip(struct cali_tc_ctx *ctx) {
-	switch (bpf_htons(ctx->skb->protocol)) {
+	__u16 protocol = 0;
+
+	if (CALI_F_XDP) {
+		if (skb_refresh_validate_ptrs(ctx, UDP_SIZE)) {
+			ctx->fwd.reason = CALI_REASON_SHORT;
+			CALI_DEBUG("Too short\n");
+			goto deny;
+		}
+		protocol = bpf_ntohs(ctx->eth->h_proto);
+	} else {
+		protocol = bpf_ntohs(ctx->skb->protocol);
```

Comment:
Originally, this line was bpf_htons(ctx->skb->protocol), however, skb.protocol type (in Kernel source: include/linux/skbuff.h) is __be16. So changed it to bpf_ntohs(ctx->skb->protocol).
Also note that protocol type in /usr/include/linux/bpf.h is __u32 (like all other fields), but I assume this is the case for alignment, and fields are mapped to kernel definition. Is that correct?

Reply:
The bpf_ntohs change LGTM.  I'm afraid I don't know why the fields in struct __sk_buff are all __u32 - @fasaxc ?